ouroboros-consensus-cardano/test/Test/ThreadNet/TxGen/Cardano.hs

```diff
@@ -1,15 +1,308 @@
+  , ctgeNetworkMagic     :: Byron.NetworkMagic
+  , ctgeShelleyCoreNodes :: [Shelley.CoreNode c]
+  }
+
+instance ( TPraosCrypto c
+           -- These equalities allow use of Byron bootstrap addresses in the
+           -- Shelley era, since @shelley-spec-ledger@ requires Ed25519 and the
+           -- current Byron-to-Shelley translation requires _224.
+         , ADDRHASH c ~ Blake2b_224
+         , DSIGN    c ~ Ed25519DSIGN
+         )
+      => TxGen (CardanoBlock c) where
+
+  type TxGenExtra (CardanoBlock c) = CardanoTxGenExtra c
+
+  -- TODO also generate " typical " Byron and Shelley transactions
+  testGenTxs (CoreNodeId i) _ncn curSlot cfg extra ls =
+      pure $ maybeToList $ migrateUTxO migrationInfo curSlot lcfg ls
+    where
+      lcfg = blockConfigLedger $ topLevelConfigBlock cfg
+
+      CardanoTxGenExtra
+        { ctgeByronGenesisKeys
+        , ctgeNetworkMagic
+        , ctgeShelleyCoreNodes
+        } = extra
+
+      GeneratedSecrets
+        { gsRichSecrets
+        } = ctgeByronGenesisKeys
+
+      migrationInfo = MigrationInfo
+        { byronMagic = ctgeNetworkMagic
+        , byronSK
+        , paymentSK
+        , poolSK
+        , stakingSK
+        , vrfSK
+        }
+
+      byronSK :: Byron.SigningKey
+      byronSK = gsRichSecrets !! fromIntegral i
+
+      Shelley.CoreNode
+        { Shelley.cnDelegateKey = paymentSK
+        , Shelley.cnStakingKey  = stakingSK
+        , Shelley.cnVRF         = vrfSK
+        } = ctgeShelleyCoreNodes !! fromIntegral i
+
+      -- Reuse the payment key as the pool key, since it's an individual
+      -- stake pool and the namespaces are separate.
+      poolSK :: SignKeyDSIGN (DSIGN c)
+      poolSK = paymentSK
+
+-- | See 'migrateUTxO'
+data MigrationInfo c = MigrationInfo
+  { byronMagic :: Byron.NetworkMagic
+    -- ^ Needed for creating a Byron address.
+  , byronSK    :: Byron.SigningKey
+    -- ^ The core node's Byron secret.
+  , paymentSK  :: SignKeyDSIGN (DSIGN c)
+  , poolSK     :: SignKeyDSIGN (DSIGN c)
+  , stakingSK  :: SignKeyDSIGN (DSIGN c)
+  , vrfSK      :: SignKeyVRF (VRF c)
+    -- ^ To be re-used by the individual pool.
+  }
+
+-- | Convert a core node's utxo from Byron to an active Shelley stake pool.
+--
+-- Returns a transaction that registers a staking key, registers an individual
+-- stake pool, delegates that stake key to that stake pool, and transfers all
+-- utxo from the Byron 'byronAddr' to the Shelley address corresponding to the
+-- pair of 'paymentSK' and 'stakingSK'.
+--
+-- It returns 'Nothing' if the core node does not have any utxo in its
+-- 'byronAddr' (eg if this transaction has already been applied).
+migrateUTxO :: forall c.
+               (TPraosCrypto c, DSIGN c ~ Ed25519DSIGN)
+            => MigrationInfo c
+            -> SlotNo
+            -> LedgerConfig (CardanoBlock c)
+            -> LedgerState (CardanoBlock c)
+            -> Maybe (GenTx (CardanoBlock c))
+migrateUTxO migrationInfo curSlot lcfg lst
```

Comment:
I must say that I'm not that familiar with all this ledger business. Best to let @nc6 review this part.

Reply:
Yeah, that's what I had in mind when I added @nc6 as a reviewer: this part and then just a general FYI.